README.md
# gstat_exporter
Prometheus exporter for FreeBSD gstat data.

# Blog post
https://blog.tyk.nu/blog/all-in-a-days-work-prometheus-gstat_exporter-and-grafana-dashboard/

# Grafana Dashboard
https://grafana.com/grafana/dashboards/11223

# Example Metrics
The following is an example of the metrics exported from a system with a small 500G flashdisk and two large spinning disks:

```
root@plads1:~ # fetch -qo - http://127.0.0.1:9248
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 426.0
python_gc_objects_collected_total{generation="1"} 0.0
python_gc_objects_collected_total{generation="2"} 0.0
# HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 48.0
python_gc_collections_total{generation="1"} 4.0
python_gc_collections_total{generation="2"} 0.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="9",patchlevel="18",version="3.9.18"} 1.0
# HELP gstat_up The value of this Gauge is always 1 when the gstat_exporter is up
# TYPE gstat_up gauge
gstat_up 1.0
# HELP gstat_queue_depth The queue depth for this GEOM
# TYPE gstat_queue_depth gauge
gstat_queue_depth{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_queue_depth{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_queue_depth{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_total_operations_per_second The total number of operations/second for this GEOM
# TYPE gstat_total_operations_per_second gauge
gstat_total_operations_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_total_operations_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_total_operations_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_read_operations_per_second The number of read operations/second for this GEOM
# TYPE gstat_read_operations_per_second gauge
gstat_read_operations_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_read_operations_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_read_operations_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_read_size_kilobytes The size in kilobytes of read operations for this GEOM
# TYPE gstat_read_size_kilobytes gauge
gstat_read_size_kilobytes{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_read_size_kilobytes{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_read_size_kilobytes{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_read_kilobytes_per_second The speed in kilobytes/second of read operations for this GEOM
# TYPE gstat_read_kilobytes_per_second gauge
gstat_read_kilobytes_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_read_kilobytes_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_read_kilobytes_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_miliseconds_per_read The speed in miliseconds/read operation for this GEOM
# TYPE gstat_miliseconds_per_read gauge
gstat_miliseconds_per_read{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_miliseconds_per_read{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_miliseconds_per_read{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_write_operations_per_second The number of write operations/second for this GEOM
# TYPE gstat_write_operations_per_second gauge
gstat_write_operations_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_write_operations_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_write_operations_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_write_size_kilobytes The size in kilobytes of write operations for this GEOM
# TYPE gstat_write_size_kilobytes gauge
gstat_write_size_kilobytes{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_write_size_kilobytes{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_write_size_kilobytes{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_write_kilobytes_per_second The speed in kilobytes/second of write operations for this GEOM
# TYPE gstat_write_kilobytes_per_second gauge
gstat_write_kilobytes_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_write_kilobytes_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_write_kilobytes_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_miliseconds_per_write The speed in miliseconds/write operation for this GEOM
# TYPE gstat_miliseconds_per_write gauge
gstat_miliseconds_per_write{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_miliseconds_per_write{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_miliseconds_per_write{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_delete_operations_per_second The number of delete operations/second for this GEOM
# TYPE gstat_delete_operations_per_second gauge
gstat_delete_operations_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_delete_operations_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_delete_operations_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_delete_size_kilobytes The size in kilobytes of delete operations for this GEOM
# TYPE gstat_delete_size_kilobytes gauge
gstat_delete_size_kilobytes{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_delete_size_kilobytes{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_delete_size_kilobytes{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_delete_kilobytes_per_second The speed in kilobytes/second of delete operations for this GEOM
# TYPE gstat_delete_kilobytes_per_second gauge
gstat_delete_kilobytes_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_delete_kilobytes_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_delete_kilobytes_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_miliseconds_per_delete The speed in miliseconds/delete operation for this GEOM
# TYPE gstat_miliseconds_per_delete gauge
gstat_miliseconds_per_delete{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_miliseconds_per_delete{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_miliseconds_per_delete{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_other_operations_per_second The number of other operations (BIO_FLUSH)/second for this GEOM
# TYPE gstat_other_operations_per_second gauge
gstat_other_operations_per_second{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_other_operations_per_second{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_other_operations_per_second{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_miliseconds_per_other The speed in miliseconds/other operation (BIO_FLUSH) for this GEOM
# TYPE gstat_miliseconds_per_other gauge
gstat_miliseconds_per_other{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_miliseconds_per_other{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_miliseconds_per_other{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
# HELP gstat_percent_busy The percent of the time this GEOM is busy
# TYPE gstat_percent_busy gauge
gstat_percent_busy{descr="WDC WDS500G2B0B-00YS70",fwheads="16",fwsectors="63",ident="212989464604",lunid="5001b444a72bc7d9",mediasize="500107862016 (466G)",name="ada0",rotationrate="0",sectorsize="512"} 0.0
gstat_percent_busy{descr="WDC WD181KRYZ-01AGBB0",fwheads="16",fwsectors="63",ident="3WJJG8NK",lunid="5000cca284e3aa8c",mediasize="18000207937536 (16T)",name="ada1",rotationrate="7200",sectorsize="512"} 0.0
gstat_percent_busy{descr="ST18000NE000-2YY101",fwheads="16",fwsectors="63",ident="ZR50A7AJ",lunid="5000c500c7dc3971",mediasize="18000207937536 (16T)",name="ada2",rotationrate="7200",sectorsize="512"} 0.0
```


CHANGELOG.md
# Changelog

All notable changes to `gstat_exporter` will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [v0.2.0] - 2023-12-14

### Added

- Added argparse and commandline arguments -p and -l to make listen ip and listen port configurable
- Added feature to remove devices when they disappear from the system
- Added -d / --debug flag to enable debug mode

### Changed
- Replace black + flake8 linters with ruff
- Refactor code into a GstatExporter class
- Made the gstat sleep period configurable with -s / --sleep and change the default from 5 secs to 15 secs (the default scrape interval in Prometheus)
- Add timestamp and configure logformat so it now looks like this:
    "2023-12-14 13-26-14 +0100 - gstat_exporter - LEVEL - message"


## [v0.1.0] - 2023-12-13

### Changed
- Create Python package
- Rename default branch from "master" to "main"


### Added

- CHANGELOG.md
- RELEASE.md


gstat_exporter.py
from prometheus_client import start_http_server, Gauge
import argparse
import logging
import datetime
from subprocess import Popen, PIPE
from typing import Dict
from importlib.metadata import PackageNotFoundError, version

# get version
try:
    __version__ = version("gstat_exporter")
except PackageNotFoundError:
    # package is not installed, version unknown
    __version__ = "0.0.0"

class GstatExporter:
    def __init__(self, interval: int = 30, grace: int = 30, sleep: int = 15) -> None:
        """Define metrics and other neccesary variables."""
        # save interval, grace, and sleep
        self.interval = interval
        self.grace = grace
        self.sleep = sleep

        # save the version as a class attribute
        self.__version__ = __version__

        # define the metric labels
        self.labels: list[str] = [
            "name",
            "descr",
            "mediasize",
            "sectorsize",
            "lunid",
            "ident",
            "rotationrate",
            "fwsectors",
            "fwheads",
        ]

        # define the metrics
        self.metrics: dict[str, Gauge] = {}
        self.metrics["up"] = Gauge(
            "gstat_up",
            "The value of this Gauge is always 1 when the gstat_exporter is up",
        )

        self.metrics["queue"] = Gauge(
            "gstat_queue_depth",
            "The queue depth for this GEOM",
            self.labels,
        )
        self.metrics["totalops"] = Gauge(
            "gstat_total_operations_per_second",
            "The total number of operations/second for this GEOM",
            self.labels,
        )

        self.metrics["readops"] = Gauge(
            "gstat_read_operations_per_second",
            "The number of read operations/second for this GEOM",
            self.labels,
        )
        self.metrics["readsize"] = Gauge(
            "gstat_read_size_kilobytes",
            "The size in kilobytes of read operations for this GEOM",
            self.labels,
        )
        self.metrics["readkbs"] = Gauge(
            "gstat_read_kilobytes_per_second",
            "The speed in kilobytes/second of read operations for this GEOM",
            self.labels,
        )
        self.metrics["readms"] = Gauge(
            "gstat_miliseconds_per_read",
            "The speed in miliseconds/read operation for this GEOM",
            self.labels,
        )

        self.metrics["writeops"] = Gauge(
            "gstat_write_operations_per_second",
            "The number of write operations/second for this GEOM",
            self.labels,
        )
        self.metrics["writesize"] = Gauge(
            "gstat_write_size_kilobytes",
            "The size in kilobytes of write operations for this GEOM",
            self.labels,
        )
        self.metrics["writekbs"] = Gauge(
            "gstat_write_kilobytes_per_second",
            "The speed in kilobytes/second of write operations for this GEOM",
            self.labels,
        )
        self.metrics["writems"] = Gauge(
            "gstat_miliseconds_per_write",
            "The speed in miliseconds/write operation for this GEOM",
            self.labels,
        )

        self.metrics["deleteops"] = Gauge(
            "gstat_delete_operations_per_second",
            "The number of delete operations/second for this GEOM",
            self.labels,
        )
        self.metrics["deletesize"] = Gauge(
            "gstat_delete_size_kilobytes",
            "The size in kilobytes of delete operations for this GEOM",
            self.labels,
        )
        self.metrics["deletekbs"] = Gauge(
            "gstat_delete_kilobytes_per_second",
            "The speed in kilobytes/second of delete operations for this GEOM",
            self.labels,
        )
        self.metrics["deletems"] = Gauge(
            "gstat_miliseconds_per_delete",
            "The speed in miliseconds/delete operation for this GEOM",
            self.labels,
        )

        self.metrics["otherops"] = Gauge(
            "gstat_other_operations_per_second",
            "The number of other operations (BIO_FLUSH)/second for this GEOM",
            self.labels,
        )
        self.metrics["otherms"] = Gauge(
            "gstat_miliseconds_per_other",
            "The speed in miliseconds/other operation (BIO_FLUSH) for this GEOM",
            self.labels,
        )

        self.metrics["busy"] = Gauge(
            "gstat_percent_busy",
            "The percent of the time this GEOM is busy",
            self.labels,
        )

        # start with an empty deviceinfo dict and add devices as we see them
        self.deviceinfo: Dict[str, Dict[str, str]] = {}

        # variables used for checking for removed devices
        self.lastcheck = datetime.datetime.now()
        self.timestamps: Dict[str, datetime.datetime] = {}

        logging.debug("Done initialising GstatExporter class")

    def get_deviceinfo(self, name: str) -> Dict[str, str]:
        """
        Return a dict of GEOM device info for GEOM devices in class DISK,
        for use as labels for the metrics.

        Sample output from the geom command:

        $ geom -p ada0
        Geom class: DISK
        Geom name: ada0
        Providers:
        1. Name: ada0
           Mediasize: 250059350016 (233G)
           Sectorsize: 512
           Mode: r2w2e4
           descr: Samsung SSD 860 EVO mSATA 250GB
           lunid: 5002538e700b753f
           ident: S41MNG0K907238X
           rotationrate: 0
           fwsectors: 63
           fwheads: 16
        $
        """
        logging.debug(f"Getting deviceinfo for GEOM {name}...")
        with Popen(
            ["geom", "-p", name], stdout=PIPE, bufsize=1, universal_newlines=True
        ) as p:
            result = {}
            for line in p.stdout:  # type: ignore
                # remove excess whitespace
                line = line.strip()
                # we only care about the DISK class for now
                if line[0:12] == "Geom class: " and line[-4:] != "DISK":
                    break

                if line[0:11] == "Mediasize: ":
                    result["mediasize"] = line[11:]
                if line[0:12] == "Sectorsize: ":
                    result["sectorsize"] = line.split(" ")[1]
                if line[0:7] == "descr: ":
                    result["descr"] = " ".join(line.split(" ")[1:])
                if line[0:7] == "lunid: ":
                    result["lunid"] = line.split(" ")[1]
                if line[0:7] == "ident: ":
                    result["ident"] = line.split(" ")[1]
                if line[0:14] == "rotationrate: ":
                    result["rotationrate"] = line.split(" ")[1]
                if line[0:11] == "fwsectors: ":
                    result["fwsectors"] = line.split(" ")[1]
                if line[0:9] == "fwheads: ":
                    result["fwheads"] = line.split(" ")[1]
            logging.debug(f"Returning deviceinfo for {name}: {result}")
            return result

    def run_gstat_forever(self) -> None:
        """
        Run gstat in a loop and update stats per line
        """
        logging.debug(f"Running 'gstat -pdosCI {self.sleep}s' (will loop forever)...")
        with Popen(
            ["gstat", "-pdosCI", f"{self.sleep}s"], stdout=PIPE, bufsize=1, universal_newlines=True
        ) as p:
            # loop over lines in the output
            for line in p.stdout:  # type: ignore
                (
                    timestamp,
                    name,
                    queue_depth,
                    total_operations_per_second,
                    read_operations_per_second,
                    read_size_kilobytes,
                    read_kilobytes_per_second,
                    miliseconds_per_read,
                    write_operations_per_second,
                    write_size_kilobytes,
                    write_kilobytes_per_second,
                    miliseconds_per_write,
                    delete_operations_per_second,
                    delete_size_kilobytes,
                    delete_kilobytes_per_second,
                    miliseconds_per_delete,
                    other_operations_per_second,
                    miliseconds_per_other,
                    percent_busy,
                ) = line.split(",")
                if timestamp == "timestamp":
                    # skip header line
                    continue

                # first check if this GEOM has been seen before
                if name not in self.deviceinfo:
                    logging.info(f"Adding new GEOM to deviceinfo: {name}")
                    # this is the first time we see this GEOM
                    self.deviceinfo[name] = {}
                    # we always need a value for all labels
                    for key in self.labels:
                        self.deviceinfo[name][key] = ""
                    # get real info from the device if it is class DISK
                    self.deviceinfo[name].update(self.get_deviceinfo(name))
                    self.deviceinfo[name].update({"name": name})

                # update timestamp to track when this GEOM was last seen
                self.timestamps[name] = datetime.datetime.strptime(
                    timestamp.split(".")[0], "%Y-%m-%d %H:%M:%S"
                )

                # up is always.. up
                self.metrics["up"].set(1)

                self.metrics["queue"].labels(**self.deviceinfo[name]).set(queue_depth)
                self.metrics["totalops"].labels(**self.deviceinfo[name]).set(
                    total_operations_per_second
                )

                self.metrics["readops"].labels(**self.deviceinfo[name]).set(
                    read_operations_per_second
                )
                self.metrics["readsize"].labels(**self.deviceinfo[name]).set(
                    read_size_kilobytes
                )
                self.metrics["readkbs"].labels(**self.deviceinfo[name]).set(
                    read_kilobytes_per_second
                )
                self.metrics["readms"].labels(**self.deviceinfo[name]).set(
                    miliseconds_per_read
                )

                self.metrics["writeops"].labels(**self.deviceinfo[name]).set(
                    write_operations_per_second
                )
                self.metrics["writesize"].labels(**self.deviceinfo[name]).set(
                    write_size_kilobytes
                )
                self.metrics["writekbs"].labels(**self.deviceinfo[name]).set(
                    write_kilobytes_per_second
                )
                self.metrics["writems"].labels(**self.deviceinfo[name]).set(
                    miliseconds_per_write
                )

                self.metrics["deleteops"].labels(**self.deviceinfo[name]).set(
                    delete_operations_per_second
                )
                self.metrics["deletesize"].labels(**self.deviceinfo[name]).set(
                    delete_size_kilobytes
                )
                self.metrics["deletekbs"].labels(**self.deviceinfo[name]).set(
                    delete_kilobytes_per_second
                )
                self.metrics["deletems"].labels(**self.deviceinfo[name]).set(
                    miliseconds_per_delete
                )

                self.metrics["otherops"].labels(**self.deviceinfo[name]).set(
                    other_operations_per_second
                )
                self.metrics["otherms"].labels(**self.deviceinfo[name]).set(
                    miliseconds_per_other
                )

                self.metrics["busy"].labels(**self.deviceinfo[name]).set(percent_busy)

                # check for removed GEOMs
                now = datetime.datetime.now()
                if (now - self.lastcheck).seconds > self.interval:
                    logging.debug("Running periodic check for removed devices...")
                    # enough time has passed since the last check
                    # loop over devices and check timestamp for each
                    remove = []
                    for name in self.deviceinfo.keys():
                        delta = (now - self.timestamps[name]).seconds
                        if delta > self.grace:
                            remove.append(name)
                            logging.info(
                                f"It has been {self.grace} seconds since gstat last reported data for GEOM {name} - removing metrics"
                            )

                    # loop over the GEOMs for which gstat stopped giving data and remove them
                    for name in remove:
                        # it has been too long since we have seen this GEOM, remove it
                        for metric in self.metrics.keys():
                            if metric == "up":
                                # skip the up metric
                                continue
                            self.metrics[metric].remove(*self.deviceinfo[name].values())
                        del self.deviceinfo[name]
                    self.lastcheck = datetime.datetime.now()


def main() -> None:
    """Run the main function."""
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-g",
        "--grace",
        type=int,
        help="Stop exporting metrics for a GEOM after gstat has not reported data from it for this many seconds. Defaults to 30 seconds.",
        default=30,
    )
    parser.add_argument(
        "-i",
        "--interval",
        type=int,
        help="How many seconds to wait between checking for removed devices. Defaults to 30 seconds.",
        default=30,
    )
    parser.add_argument(
        "-l",
        "--listen-ip",
        type=str,
        help="Listen IP. Defaults to 0.0.0.0 (all v4 IPs). Set to :: to listen on all v6 IPs.",
        default="0.0.0.0",
    )
    parser.add_argument(
        "-p",
        "--port",
        type=int,
        help="Portnumber. Defaults to 9248.",
        default=9248,
    )
    parser.add_argument(
        "-s",
        "--sleep",
        type=int,
        help="How long should gstat sleep between reporting data, in seconds. Set this to the same as your Prometheus scrape interval. Defaults to 15.",
        default=15,
    )

    parser.add_argument(
        "-d",
        "--debug",
        action="store_const",
        dest="loglevel",
        const="DEBUG",
        help="Debug mode.",
        default="INFO",
    )

    args = parser.parse_args()
    logging.basicConfig(level=args.loglevel, datefmt="%Y-%m-%d %H:%M:%S %z", format="%(asctime)s - %(module)s - %(levelname)s - %(message)s")
    logging.info(f"Starting gstat_exporter v{__version__} - logging at level {args.loglevel}")
    logging.info(f"Starting HTTP listener on address '{args.listen_ip}' port '{args.port}'")
    start_http_server(addr=args.listen_ip, port=args.port)
    exporter = GstatExporter(interval=args.interval, grace=args.grace, sleep=args.sleep)
    while True:
        exporter.run_gstat_forever()

if __name__ == "__main__":
    main()
